### PR TITLE
Lock device when computer is locked, sleeping, shutting down, or logging off

### PR DESCRIPTION
**Pull request by netromdk**
I've implemented a `SystemEventHandler` that currently listens for screen lock system events on Mac and Windows. It then emits `screenLocked()` on that occasion.

The main window UI was extended with a check box for enabling this behavior. When `screenLocked()` is caught it will call `WSClient::sendLockDevice()` that sends JSON and in `WSServerCon` calls
```cpp
mpdevice->sendData(MPCmd::Command::LOCK_DEVICE, cb);
```

Toggling the check box will set/change a bool via `QSettings` for the key `"settings/LockDeviceOnSystemEvents"`, which is used for the state when MC starts.

As @raoulh mentioned, I waited with Linux support until this feature has matured.

I also added events for system sleep, shutdown, and user log off.

### CODE REVIEW

File: src/MacSystemEvents.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#ifndef SRC_MACSYSTEMEVENTS_H
+#define SRC_MACSYSTEMEVENTS_H
+
+#ifdef __cplusplus
```
Comment (raoulh) [inline src/MacSystemEvents.h]: Are those guard really needed?

File: src/MainWindow.cpp [hunk 2]
```diff
@@ -492,6 +492,14 @@ MainWindow::MainWindow(WSClient *client, DbMasterController *mc, QWidget *parent
     connect(ui->checkBoxBoot, SIGNAL(toggled(bool)), this, SLOT(checkSettingsChanged()));
     connect(ui->checkBoxTuto, SIGNAL(toggled(bool)), this, SLOT(checkSettingsChanged()));
 
+#ifdef Q_OS_LINUX
+    ui->checkBoxLockDevice->setChecked(false);
+    ui->checkBoxLockDevice->setEnabled(false);
```
Comment (raoulh) [inline src/MainWindow.cpp]: I would prefer to hide the checkbox on linux for now

File: src/MainWindow.cpp [hunk 3]
```diff
@@ -508,6 +516,17 @@ MainWindow::MainWindow(WSClient *client, DbMasterController *mc, QWidget *parent
     //Setup the confirm view
     ui->widgetSpin->setPixmap(AppGui::qtAwesome()->icon(fa::circleonotch).pixmap(QSize(80, 80)));
 
+#if defined(Q_OS_MAC) || defined(Q_OS_WIN)
+    connect(&eventHandler, &SystemEventHandler::screenLocked, this, [this]
+    {
+        const bool exec = ui->checkBoxLockDevice->isChecked();
+        if (exec && wsClient->get_status() == Common::Unlocked) {
```
Comment (raoulh) [inline src/MainWindow.cpp]: { on newline please

File: src/SystemEvent.h [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#ifndef SRC_SYSTEMEVENT_H
+#define SRC_SYSTEMEVENT_H
+
+enum SystemEvent
```
Comment (raoulh) [inline src/SystemEvent.h]: Could you move this single value enum in SystemEventHandler.h, having a .h only for 1 enum is not so nice

File: src/WSServerCon.cpp [hunk 5]
```diff
@@ -596,6 +596,11 @@ void WSServerCon::processMessage(const QString &message)
         }
         );
     }
+    else if (root["msg"] == "lock_device")
+    {
+        const auto cb = [](bool, const QByteArray &, bool &){};
+        mpdevice->sendData(MPCmd::Command::LOCK_DEVICE, cb);
```
Comment (raoulh) [inline src/WSServerCon.cpp]: This is not what you want to do. You need to add a new function in MPDevice that will create a new AsyncJob and put it onto the job queue. Directly sending a data packet like that will bypass the current running queue of command and can mess with the communication at some point

File: src/SystemEvent.h (hunk 4 above)
Comment (netromdk) [reply src/SystemEvent.h]: I have another branch which adds three values to this enum: `LOGGING_OFF`, `GOING_TO_SLEEP`, and `SHUTTING_DOWN`.

File: src/MacSystemEvents.h (hunk 1 above)
Comment (netromdk) [reply src/MacSystemEvents.h]: You're right, I actually don't. :)

File: src/MainWindow.cpp (hunk 2 above)
Comment (netromdk) [reply src/MainWindow.cpp]: I can do that.

File: src/WSServerCon.cpp (hunk 5 above)
Comment (netromdk) [reply src/WSServerCon.cpp]: That makes sense. Will change it into that approach.

**Comment by netromdk** [comment]
@raoulh Thanks for the review. I've made changes in relation to your comments.

**Comment by netromdk** [comment]
@raoulh btw I don't like the timer approach either but the Windows API deprecated and removed the [WM_WTSSESSION_CHANGE](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383828(v=vs.85).aspx) (with `wParam==WTS_SESSION_LOCK`) after Windows 2000. I tried it out anyway and it indeed doesn't get fired.

File: src/WSServerCon.cpp [hunk 6]
```diff
@@ -596,6 +596,19 @@ void WSServerCon::processMessage(const QString &message)
         }
         );
     }
+    else if (root["msg"] == "lock_device")
+    {
+        mpdevice->lockDevice([this, root](bool success, QString errstr)
+        {
+            if (!success) return;
```
Comment (raoulh) [inline src/WSServerCon.cpp]: if !success you need to send back sendFailedJson(root, errstr); before returning.

File: src/WSServerCon.cpp [hunk 7]
```diff
@@ -596,6 +596,19 @@ void WSServerCon::processMessage(const QString &message)
...
+            QJsonObject ores;
+            QJsonObject oroot = root;
+            ores["success"] = "false";
```
Comment (raoulh) [inline src/WSServerCon.cpp]: here it should be = "true"

File: src/WSServerCon.cpp [hunk 8]
```diff
@@ -596,6 +596,19 @@ void WSServerCon::processMessage(const QString &message)
...
+            ores["errstr"] = errstr;
```
Comment (raoulh) [inline src/WSServerCon.cpp]: no errstr because we are in the success case

File: src/WSServerCon.cpp (hunk 6 above)
Comment (netromdk) [reply src/WSServerCon.cpp]: The point was the if it works then it's evident by the fact that the device's status changes to locked. But sure, can send back response in every case.

File: src/WSServerCon.cpp (hunk 7 above)
Comment (netromdk) [reply src/WSServerCon.cpp]: Not when it is not a success.

File: src/WSServerCon.cpp (hunk 6 above)
Comment (netromdk) [reply src/WSServerCon.cpp]: Haha.. I actually meant `if (success) return`..

**Pull request merged**